src/array.jl

```diff
@@ -117,10 +113,16 @@ function CategoricalMatrix end
 CategoricalArray(dims::Int...; ordered=false) =
     CategoricalArray{String}(dims, ordered=ordered)
 
-CategoricalArray{T, N, R}(dims::NTuple{N,Int}; ordered=false) where {T, N, R} =
-    CategoricalArray{T, N, R}(zeros(R, dims), CategoricalPool{T, R}(ordered))
+function CategoricalArray{T, N, R}(dims::NTuple{N,Int};
+                                   ordered=false) where {T, N, R}
+    C = catvalue_type(unwrap_catvalue_type(T), R)
+    V = valtype(C)
+    S = T >: Null ? Union{V, Null} : V
+    CategoricalArray{S, N}(zeros(R, dims), CategoricalPool{V, R, C}(ordered))
```

Comment:
I don't understand what this does. It doesn't sound correct to create a `CategoricalArray{S, N}` when the caller explicitly requested a `CategoricalArray{T, N, R}`. Transformations of type parameters should only happen when they are obtained implicitly from the types of the arguments.

The job of unwrapping `CategoricalValue` eltypes should rather be done by higher-level constructors as was the case before.

Reply:
So what it exactly does:
 * `unwrap_catvalue_type(T)` gets the original type if `T` is categorical value
 * `C = catvalue_type()` selects the appropriate categorical value implementation for the original type
 * `V` is the actual value type that `C` encodes, `V` might be different from `Nulls.T(unwrap_catvalue_type(T))`, e.g. when `T == SubString`, but `convert(V, x::T)` should always be possible. As I noted in the comment the possibility `V != Nulls.T(unwrap_catvalue_type(T))` was introduced to handle `AbstractString` values.
 * `S` restores the nullability

It's true that it might be better to do type logic somewhere upstream.
But the same critics applies to ctors like `CategoricalArray{CategoricalValue{T}, N, R}` from the original code. So to keep the already substantial changes minimal I would rather keep it.

Essentially, before the PR the logic of type wrapping/transformations and handling nulls was defined by the "higher-level" ctors.
With `CategoricalValue` being split into multiple types it becomes harder to continue to do it this way. And these ctors still require the overhead of passing the normal parameters so it becomes harder to read and maintain.
So in this PR the logic of type wrapping/unwrapping is delegated to trait-based `unwrap_catvalue_type(T)`/`catvalue_type()` methods, and the "central" ctor that we are now discussing is the only place where these methods are called.
IMO it leads to leaner code.

Reply:
Hmm, let me think about this. That need to special-case `CatValue` wrapping a `CatValue` is really not pretty in the current code either.

Reply:
OK, I guess it's fine, especially if it allows removing higher-level constructors. I'm not sure I wrote it in the best way in the first place since I remember it was quite messy.